source/agora/common/Block.d

```diff
@@ -307,9 +307,25 @@ public struct Block
     {
         assert(this.merkle_tree.length != 0, "Block hasn't been fully initialized");
 
-        foreach (idx; 0 .. min(this.txs.length, this.merkle_tree.length))
-            if (hash == this.merkle_tree[idx])
-                return idx;
+        const size_t size = min(this.txs.length, this.merkle_tree.length);
+        if (size == 0) return this.txs.length;
+        long left = 0;
+        long right = size - 1;
+        long mid;
+        while (left <= right)
+        {
+            mid = (left + right) / 2;
+
+            if (this.merkle_tree[mid] == hash)
```

Comment:
you have a double lookup here. How about storing it in a temporary:

```D
auto tgt_hash = this.merkle_tree[mid];
if (tgt_hash == hash)
   ...
else
    if (hash < tgt_hash)
...
```

Reply:
Modified~ Thanks~